proposals/nnnn-custom-actor-executors.md

```diff
@@ -0,0 +1,593 @@
+# Custom Actor Executors
+
+- Proposal: [SE-NNNN](https://github.com/apple/swift-evolution/blob/custom-actor-executors/proposals/NNNN-custom-actor-executors.md)
+- Authors: [Konrad 'ktoso' Malawski](https://github.com/ktoso), [John McCall](https://github.com/rjmccall)
+- Review Manager: TBD
+- Status: **Partially implemented on `main`**
+- Previous threads:
+  - Original pitch thread from around Swift 5.5: [Support custom executors in Swift Concurrency](https://forums.swift.org/t/support-custom-executors-in-swift-concurrency/44425)
+
+## Table of Contents
+
+- [Custom Actor Executors](#custom-actor-executors)
+  * [Introduction](#introduction)
+  * [Motivation](#motivation)
+  * [Proposed solution](#proposed-solution)
+  * [Detailed design](#detailed-design)
+    + [A low-level design](#a-low-level-design)
+    + [Executors](#executors)
+    + [Serial Executors](#serial-executors)
+    + [Jobs](#jobs)
+    + [Actors with custom SerialExecutors](#actors-with-custom-serialexecutors)
+    + [Executor assertions](#executor-assertions)
+    + [Default Swift Runtime Executors](#default-swift-runtime-executors)
+  * [Source compatibility](#source-compatibility)
+  * [Effect on ABI stability](#effect-on-abi-stability)
+  * [Effect on API resilience](#effect-on-api-resilience)
+  * [Alternatives considered](#alternatives-considered)
+  * [Future Directions](#future-directions)
+    + [Overriding the MainActor executor](#overriding-the-mainactor-executor)
+    + [Executor Switching](#executor-switching)
+    + [Specifying Task executors](#specifying-task-executors)
+    + [DelegateActor property](#delegateactor-property)
+
+## Introduction
+
+As Swift Concurrency continues to mature it is becoming increasingly important to offer adopters tighter control over where exactly asynchronous work is actually executed.
+
+This proposal introduces a basic mechanism for customizing actor  executors. By providing an instance of an executor, actors can influence "where" they will be executing any task they are running, while upholding the mutial excusion and actor isolation guaranteed by the actor model.
+
+>  **Note:** This proposal defines only a set of APIs to customize actor executors, and other kinds of executor control is out of scope for this specific proposal.
+
+## Motivation
+
+Swift's concurrency design is intentionally vague about the details of how code is actually run. Most code does not rely on specific properties of the execution environment, such as being run to a specific operating system thread, and instead needs only high-level semantic properties, expressed in terms of actor isolation, such as that no other code will be accessing certain variables concurrently. Maintaining flexibility about how work is scheduled onto threads allows Swift to avoid certain performance pitfalls by default.
```

Comment:
```suggestion
Swift's concurrency design is intentionally vague about the details of how code is actually run. Most code does not rely on specific properties of the execution environment, such as being run on a specific operating system thread, and instead needs only high-level semantic properties, expressed in terms of actor isolation, such as that no other code will be accessing certain variables concurrently. Maintaining flexibility about how work is scheduled onto threads allows Swift to avoid certain performance pitfalls by default.
```